// @commitlint/cz-commitlint/lib/store/prompts.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPromptSettings = exports.getPromptQuestions = exports.getPromptMessages = exports.setPromptConfig = void 0;
const lodash_isplainobject_1 = __importDefault(require("lodash.isplainobject"));
const defaultPromptConfigs_1 = __importDefault(require("./defaultPromptConfigs"));
const storeKey = Symbol('promptConfig');
const store = {
    [storeKey]: defaultPromptConfigs_1.default,
};
function setPromptConfig(newPromptConfig) {
    const { settings, messages, questions } = newPromptConfig;
    if (messages) {
        const requiredMessageKeys = Object.keys(defaultPromptConfigs_1.default.messages);
        requiredMessageKeys.forEach((key) => {
            const message = messages[key];
            if (typeof message === 'string') {
                store[storeKey]['messages'][key] = message;
            }
        });
    }
    if (questions && (0, lodash_isplainobject_1.default)(questions)) {
        store[storeKey]['questions'] = questions;
    }
    if (settings && (0, lodash_isplainobject_1.default)(settings)) {
        if (settings['scopeEnumSeparator'] &&
            !/^\/|\\|,$/.test(settings['scopeEnumSeparator'])) {
            console.log(`prompt.settings.scopeEnumSeparator must be one of ',', '\\', '/'.`);
            process.exit(1);
        }
        store[storeKey]['settings'] = Object.assign(Object.assign({}, defaultPromptConfigs_1.default.settings), settings);
    }
}
exports.setPromptConfig = setPromptConfig;
function getPromptMessages() {
    var _a;
    return (_a = (store[storeKey] && store[storeKey]['messages'])) !== null && _a !== void 0 ? _a : {};
}
exports.getPromptMessages = getPromptMessages;
function getPromptQuestions() {
    var _a;
    return (_a = (store[storeKey] && store[storeKey]['questions'])) !== null && _a !== void 0 ? _a : {};
}
exports.getPromptQuestions = getPromptQuestions;
function getPromptSettings() {
    var _a;
    return (_a = (store[storeKey] && store[storeKey]['settings'])) !== null && _a !== void 0 ? _a : {};
}
exports.getPromptSettings = getPromptSettings;
