// @commitlint/cz-commitlint/lib/SectionHeader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getQuestionConfig = exports.getQuestions = exports.combineCommitMessage = exports.HeaderQuestion = void 0;
const Question_1 = __importDefault(require("./Question"));
const getRuleQuestionConfig_1 = __importDefault(require("./services/getRuleQuestionConfig"));
const prompts_1 = require("./store/prompts");
class HeaderQuestion extends Question_1.default {
    constructor(name, questionConfig, headerMaxLength, headerMinLength) {
        super(name, questionConfig);
        this.headerMaxLength = headerMaxLength !== null && headerMaxLength !== void 0 ? headerMaxLength : Infinity;
        this.headerMinLength = headerMinLength !== null && headerMinLength !== void 0 ? headerMinLength : 0;
    }
    beforeQuestionStart(answers) {
        const headerRemainLength = this.headerMaxLength - combineCommitMessage(answers).length;
        this.maxLength = Math.min(this.maxLength, headerRemainLength);
        this.minLength = Math.min(this.minLength, this.headerMinLength);
    }
}
exports.HeaderQuestion = HeaderQuestion;
function combineCommitMessage(answers) {
    const { type = '', scope = '', subject = '' } = answers;
    const prefix = `${type}${scope ? `(${scope})` : ''}`;
    if (subject) {
        return ((prefix ? prefix + ': ' : '') + subject).trim();
    }
    else {
        return prefix.trim();
    }
}
exports.combineCommitMessage = combineCommitMessage;
function getQuestions() {
    // header: type, scope, subject
    const questions = [];
    const headerRuleFields = ['type', 'scope', 'subject'];
    const headerRuleQuestionConfig = (0, getRuleQuestionConfig_1.default)('header');
    if (!headerRuleQuestionConfig) {
        return [];
    }
    headerRuleFields.forEach((name) => {
        const questionConfig = getQuestionConfig(name);
        if (questionConfig) {
            const instance = new HeaderQuestion(name, questionConfig, headerRuleQuestionConfig.maxLength, headerRuleQuestionConfig.minLength);
            questions.push(instance.question);
        }
    });
    return questions;
}
exports.getQuestions = getQuestions;
function getQuestionConfig(name) {
    const questionConfig = (0, getRuleQuestionConfig_1.default)(name);
    if (questionConfig) {
        if (name === 'scope') {
            if ((0, prompts_1.getPromptSettings)()['enableMultipleScopes']) {
                questionConfig.multipleSelectDefaultDelimiter =
                    (0, prompts_1.getPromptSettings)()['scopeEnumSeparator'];
            }
            // split scope string to segments, match commitlint rules
            questionConfig.multipleValueDelimiters = /\/|\\|,/g;
        }
    }
    return questionConfig;
}
exports.getQuestionConfig = getQuestionConfig;
